crates/neon/src/types_impl/function/mod.rs

```diff
@@ -7,11 +7,75 @@ use crate::{
     handle::Handle,
     object::Object,
     result::{JsResult, NeonResult},
-    types::{JsFunction, JsObject, JsValue, Value},
+    types::{
+        extract::{TryFromJs, TryIntoJs},
+        JsFunction, JsObject, JsValue, Value,
+    },
 };
 
 pub(crate) mod private;
 
+/// A builder for making a JavaScript function call like `parseInt("42")`.
+///
+/// The builder methods make it convenient to assemble the call from parts:
+/// ```
+/// # use neon::prelude::*;
+/// # fn foo(mut cx: FunctionContext) -> JsResult<JsNumber> {
+/// # let parse_int: Handle<JsFunction> = cx.global("parseInt")?;
+/// let x: f64 = parse_int
+///     .bind(&mut cx)
+///     .arg("42")?
+///     .apply()?;
+/// # Ok(cx.number(x))
+/// # }
+/// ```
+pub struct BindOptions<'a, 'cx: 'a, C: Context<'cx>> {
+    pub(crate) cx: &'a mut C,
+    pub(crate) callee: Handle<'cx, JsFunction>,
+    pub(crate) this: Option<Handle<'cx, JsValue>>,
+    pub(crate) args: private::ArgsVec<'cx>,
+}
+
+impl<'a, 'cx: 'a, C: Context<'cx>> BindOptions<'a, 'cx, C> {
+    /// Set the value of `this` for the function call.
+    pub fn this<V: Value>(&mut self, this: Handle<'cx, V>) -> &mut Self {
+        self.this = Some(this.upcast());
+        self
+    }
+
+    /// Replaces the arguments list with the given arguments.
+    pub fn args<A: TryIntoArguments<'cx>>(&mut self, a: A) -> NeonResult<&mut Self> {
+        self.args = a.try_into_args_vec(self.cx)?;
+        Ok(self)
+    }
+
+    /// Add an argument to the arguments list.
+    pub fn arg<A: TryIntoJs<'cx>>(&mut self, a: A) -> NeonResult<&mut Self> {
+        let v = a.try_into_js(self.cx)?;
+        self.args.push(v.upcast());
+        Ok(self)
+    }
+
+    /// Add an argument to the arguments list, computed from a closure.
+    pub fn arg_with<R, F>(&mut self, f: F) -> NeonResult<&mut Self>
+    where
+        R: TryIntoJs<'cx>,
+        F: FnOnce(&mut C) -> NeonResult<R>,
+    {
+        let v = f(self.cx)?.try_into_js(self.cx)?;
+        self.args.push(v.upcast());
+        Ok(self)
+    }
+
+    /// Make the function call. If the function returns without throwing, the result value
+    /// is converted to a Rust value with `TryFromJs::from_js`.
+    pub fn apply<R: TryFromJs<'cx>>(&mut self) -> NeonResult<R> {
+        let this = self.this.unwrap_or_else(|| self.cx.undefined().upcast());
+        let v: Handle<JsValue> = self.callee.call(self.cx, this, &self.args)?;
+        R::from_js(self.cx, v)
```

Comment:
We haven't published `TryFromJs` yet and can still make breaking changes. This makes me wonder if we *shouldn't* make `from_js` part of the API and should _instead_ create a `impl TryFromJs for NeonResult`. 🤔 

We can flag this for follow-up.

Reply:
Nevermind, we can't actually have this. It would end up with a `NeonResult<Result<NeonResult<T>, Infallible>>`, which is really messy.

I think what I could do is make `from_js` a different trait with a blanket implementation`where <Self as TryFromJs<'cx>>::Error: TryIntoJs<'cx>`. We can discuss that outside this PR.